src/backend/distributed/commands/table.c

```diff
@@ -593,59 +609,185 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand)
+
+	/*
+	 * To prevent deadlocks, sort the list before converting each postgres local
+	 * table to a citus local table.
+	 */
+	relationRangeVarList = SortList(relationRangeVarList, CompareRangeVarsByOid);
+
+	/*
+	 * Here we should operate on RangeVar objects since relations oid's would
+	 * change in below loop due to CreateCitusLocalTable.
+	 */
+	RangeVar *relationRangeVar;
+	foreach_ptr(relationRangeVar, relationRangeVarList)
+	{
+		List *commandList = alterTableStatement->cmds;
+		LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
+		bool missingOk = alterTableStatement->missing_ok;
+		Oid rightRelationId = RangeVarGetRelid(relationRangeVar, lockMode, missingOk);
+		if (!OidIsValid(rightRelationId))
+		{
+			/*
+			 * As we are in preprocess, missingOk might be true and relation
+			 * might not exist.
+			 */
+			continue;
+		}
+		else if (IsCitusTable(rightRelationId))
 		{
-			ErrorOutForFKeyBetweenPostgresAndCitusLocalTable(leftRelationId);
+			/*
+			 * relationRangeVarList has also reference and citus local tables
+			 * involved in this ADD FOREIGN KEY command. Morever, even if
+			 * rightRelationId was belonging to a postgres  local table initially,
+			 * we might had already converted it to a citus local table by cascading.
+			 */
+			continue;
 		}
+
+		/* TODO: should use try/catch block for CreateCitusLocalTable function */
+		bool cascade = true;
+		CreateCitusLocalTable(rightRelationId, cascade);
 	}
 }
 
 
 /*
- * GetAlterTableStmtFKeyConstraintList returns a list of Constraint objects for
- * the foreign keys that given ALTER TABLE statement defines.
+ * CompareRangeVarsByOid is a comparison function to sort RangeVar object list.
+ */
+static int
+CompareRangeVarsByOid(const void *leftElement, const void *rightElement)
+{
+	RangeVar *leftRangeVar = *((RangeVar **) leftElement);
+	RangeVar *rightRangeVar = *((RangeVar **) rightElement);
+
+	/*
+	 * Any way we will check their existence, so it's okay to map non-existing
+	 * relations to InvalidOid when sorting.
+	 */
+	bool missingOk = true;
+
+	/*
+	 * As this is an object comparator function, there is no way to understand
+	 * proper lock mode. So assume caller already locked relations.
+	 */
+	LOCKMODE lockMode = NoLock;
+
+	Oid leftRelationId = RangeVarGetRelid(leftRangeVar, lockMode, missingOk);
+	Oid rightRelationId = RangeVarGetRelid(rightRangeVar, lockMode, missingOk);
+	return CompareOids(&leftRelationId, &rightRelationId);
+}
+
+
+/*
+ * GetAlterTableAddFKeyRightRelationIdList returns a list of oid's for right
+ * relations involved in foreign keys to be defined by given ALTER TABLE command.
+ */
+static List *
+GetAlterTableAddFKeyRightRelationIdList(AlterTableStmt *alterTableStatement)
+{
+	List *rightRelationRangeVarList =
+		GetAlterTableAddFKeyRightRelationRangeVarList(alterTableStatement);
+	List *commandList = alterTableStatement->cmds;
+	LOCKMODE lockMode = AlterTableGetLockLevel(commandList);
```

Comment:
I guess we should make this a little more explicit, iterating over `commandList ` and only acting on `FOREIGN KEY`s

Reply:
Indeed this is what we do by calling `GetAlterTableAddFKeyRightRelationRangeVarList -> GetAlterTableAddFKeyConstraintList -> GetAlterTableCommandFKeyConstraintList`

Reply:
I think what confuses me a lot is that we check the `AlterTableStmt` command type too late/too deep, inside `GetAlterTableCommandFKeyConstraintList `. Maybe in terms of correctness it could be OK, but readability/simplicity wise we could improve. 

Can we do something like the following, and use `l` in as a param to the other functions?
```C
static void
ConvertPostgresLocalTablesToCitusLocalTables(AlterTableStmt *alterTableStatement)
{
	List *l = GetAlterTableAddFKeyConstraintList(alterTableStatement);
	if (l == NIL)
		return;
```

Does that sound good to you?

Reply:
I'm okay with doing this check beforehand:

>	List *l = GetAlterTableAddFKeyConstraintList(alterTableStatement);
	if (l == NIL)
		return false;

but passing this list to below calls might decrease the readability I think, what do you think ?